### DISCUSSION TITLE
Plans for metric-based autoscaling

### DISCUSSION THREAD

**Discussion by tomkerkhove** [Azure/bellhop#59] [Q&A]
Are there plans to provide metric-based autoscaling or purely time-based autoscaling?

**Comment by CloudViking** [comment]
Thank you for opening a discussion around this topic, we appreciate the engagement!  We do have plans to investigate this topic, currently open as issue #9. I think this is an interesting idea but definitely see how this would make Bellhop much more complicated. My initial thought around a potential solution for this would be to create a new Engine Function:

- 2 separate Engine Functions
    - Because each Azure Function can have only 1 trigger

- As part of deployment the user could choose to deploy either option or both:
    - Event-Hub triggered Azure Function (New Function)
    - Timer Trigger Azure Function (Existing Function)

- Both Functions would write the same format message to a single storage queue
- Scaler functions/modules theoretically should not change

Definitely open to ideas and potential use cases!

**Reply by tomkerkhove** [reply]
Interesting approach with Event Hubs - What would be feeding it then?

Based on my experiment to [autoscale ACI](https://github.com/tomkerkhove/autoscale-aci-with-azure-serverless), I can highly recommend to rely on Azure Monitor and alerts to do the decision making.

That could then trigger your function and act as a scale controller to orchestrate the required actions.

Is the engine or scaler in charge of determining what to scale to? The approach I used was to let it up to the scale controller and just give the scalers all they need. (But need to check the Bellhop details, but we can't rely on tags here)

In hindsight of my experiment, Azure Durable Entities could make a good scale controller.

Happy to help where I can though, just thinking out loud.

**Comment by CloudViking** [comment]
My idea was to still feed Event Hubs with Azure monitor data, but use diagnostic settings to send to Event Hubs. Thinking through this that is probably overkill and triggering the function directly from the Azure Monitor action group is probably cleanest. 

In the current design, the end-user is in control of which tier to scale their resources to. The Engine determines which direction to scale the resource, and the Scalers execute that operation. For all of the currently supported resources, the `setState-` value is used to determine the "tier" to scale DOWN to. When scaling UP Bellhop uses the `saveState-` tags which are applied by the Scaler based on the Graph API query result it receives from the Storage Queue.

I'm not seeing a way to allow for scheduled and alert-based scaling via a single Engine function.

We appreciate the ideas and discussion!

**Reply by tomkerkhove** [reply]
Yes, I think those have to be split indeed, but there's also another aspect which is - Should metric-based stick with scale up/down or does it scale out/in? But this could be seen as an extension as scale in/out can be relevant for time-based scaling as well (ie for #58 and #56)

In terms of the engine, I agree that it needs to be split.

This is how I've approached it:

1. **Azure Monitor Alerts are used to monitor scaling criteria** and will trigger scale controller to make scaling decisions. Given they don't automatically mitigate they will keep on sending notifications until the scale criteria is no longer applicable.
2. **Scale controllers evaluate the current amount of replicas** when they are triggered and take max/min replicas into account. In case scaling actions have to occur, they trigger a dedicated workflow to make it so and wait until it completes to trigger a scaling event.
3. **Scale In** workflow queries the current amount of Azure Container Instance container group in the resource group and delete the first one.
4. **Scale Out** workflow will load a deployment template provided in Azure Storage and replace it with the required information for the new container group. After that, it fully relies on Azure Resource Manager (ARM).
5. **Azure Resource Manager provisions a new Azure Container Instance container group** based on the provided template in Azure Storage. When required, it can automatically add an endpoint in Azure Traffic Manager to expose it to clients.

![image](https://user-images.githubusercontent.com/4345663/114132924-84fd2180-9905-11eb-94a9-a067a8797fc8.png)

Bringing this back to Bellhop, I see the engine as the scale controller and the scalers are what the scale in/out controllers are doing.

The difference, though, is that the scalers are very dumb and only doing what the scale controller is asking it to do. So that would be a (big) change for Bellhop, but that allows you to make all the decisions in the controller and let the scaler just scale.

If you want to do scale up/down/in/out, the engine knows what to do and ask the scaler.

Another thing to keep in mind is the queue when interacting with scalers. It would need to use expiry of the commands to avoid that the scaling is happening 15 min later while it's no longer required for example, but that's already going in a not of details and (premature) optimization.
